File: sendToPython/app/src/main/java/com/lullulalal/sendtopython/MainActivity.java
package com.lullulalal.sendtopython;

import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import android.Manifest;

import androidx.activity.EdgeToEdge;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.camera.core.Camera;
import androidx.camera.core.CameraSelector;
import androidx.camera.core.ImageCapture;
import androidx.camera.core.ImageCaptureException;
import androidx.camera.core.Preview;
import androidx.camera.lifecycle.ProcessCameraProvider;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.camera.view.PreviewView;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.InputStream;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import com.chaquo.python.PyObject;
import com.chaquo.python.Python;
import com.chaquo.python.android.AndroidPlatform;
import com.google.common.util.concurrent.ListenableFuture;

import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class MainActivity extends AppCompatActivity {

    private PreviewView previewView;
    private ImageCapture imageCapture;
    private ExecutorService cameraExecutor;
    private static final int REQUEST_CAMERA_PERMISSION = 100;

    private final ActivityResultLauncher<String> requestPermissionLauncher =
            registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
                if (isGranted) {
                    startCamera();
                } else {
                    Toast.makeText(this, "카메라 권한이 필요합니다.", Toast.LENGTH_SHORT).show();
                }
            });

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_main);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        //파이썬 초기화
        if (!Python.isStarted()) {
            Python.start(new AndroidPlatform(this));
        }

        Python py = Python.getInstance();
        PyObject module = py.getModule("mymodule");

//        PyObject resultObj = module.callAttr("add",12,30);
//
//        // 결과값을 Java String 또는 int 등으로 변환
//        int sum = resultObj.toInt();
//        String resultText = String.valueOf(sum); // int를 String으로 변환
//
//        // TextView에 결과 설정
//        tvResult.setText(resultText);
//
//        PyObject resultObj2 = module.callAttr("add_and_sum",55,41);
//        java.util.List<PyObject> tupleAsList = resultObj2.asList();
//
//        int sum2 = tupleAsList.get(0).toInt(); // 첫 번째 요소 (합)
//        int diff = tupleAsList.get(1).toInt(); // 두 번째 요소 (차)
//
//        Log.d("Python","Sum:"+sum2+" diff:"+diff); //Sum:96 diff:14

//        // 사진
//        PyObject result = module.callAttr("capture_and_get_bytes");
//        if (result != null && !result.isEmpty()) {
//            byte[] imageBytes = result.toJava(byte[].class);
//
//            Bitmap bitmap = BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.length);
//            iv.setImageBitmap(bitmap);
//        } else {
//            Log.d("TAG", "Failed to capture or encode image from Python");
//        }

        previewView = findViewById(R.id.previewView);
        Button captureButton = findViewById(R.id.captureButton);

        requestPermissionLauncher.launch(Manifest.permission.CAMERA);

        captureButton.setOnClickListener(v -> takePhoto());

        cameraExecutor = Executors.newSingleThreadExecutor();
    }

    private void startCamera() {
        ListenableFuture<ProcessCameraProvider> cameraProviderFuture =
                ProcessCameraProvider.getInstance(this);

        cameraProviderFuture.addListener(() -> {
            try {
                ProcessCameraProvider cameraProvider = cameraProviderFuture.get();
                Preview preview = new Preview.Builder().build();
                preview.setSurfaceProvider(previewView.getSurfaceProvider());

                imageCapture = new ImageCapture.Builder().build();

                CameraSelector cameraSelector = CameraSelector.DEFAULT_BACK_CAMERA;

                cameraProvider.unbindAll();
                cameraProvider.bindToLifecycle(this, cameraSelector, preview, imageCapture);

            } catch (ExecutionException | InterruptedException e) {
                e.printStackTrace();
            }
        }, ContextCompat.getMainExecutor(this));
    }

    private void takePhoto() {
        if (imageCapture == null) return;

        File photoFile = new File(getExternalFilesDir(null), "photo.jpg");

        ImageCapture.OutputFileOptions outputOptions =
                new ImageCapture.OutputFileOptions.Builder(photoFile).build();

        imageCapture.takePicture(
                outputOptions,
                ContextCompat.getMainExecutor(this),
                new ImageCapture.OnImageSavedCallback() {
                    @Override
                    public void onImageSaved(@NonNull ImageCapture.OutputFileResults outputFileResults) {
                        Toast.makeText(MainActivity.this, "사진 저장됨", Toast.LENGTH_SHORT).show();
                        uploadImage(photoFile);
                    }

                    @Override
                    public void onError(@NonNull ImageCaptureException exception) {
                        exception.printStackTrace();
                    }
                }
        );
    }

    private void uploadImage(File file) {
        new Thread(() -> {
            try {
                OkHttpClient client = new OkHttpClient();

                InputStream inputStream = new FileInputStream(file);
                byte[] bytes = getBytes(inputStream);

                RequestBody requestBody = new MultipartBody.Builder()
                        .setType(MultipartBody.FORM)
                        .addFormDataPart("file", "photo.jpg",
                                RequestBody.create(MediaType.parse("image/jpeg"), bytes))
                        .build();

                Request request = new Request.Builder()
                        .url("http://192.168.219.115:5000/upload")  // Flask 서버 주소
                        .post(requestBody)
                        .build();

                Response response = client.newCall(request).execute();

                if (response.isSuccessful()) {
                    runOnUiThread(() -> Toast.makeText(MainActivity.this, "업로드 성공", Toast.LENGTH_SHORT).show());
                } else {
                    runOnUiThread(() -> Toast.makeText(MainActivity.this, "업로드 실패", Toast.LENGTH_SHORT).show());
                }

            } catch (Exception e) {
                e.printStackTrace();
            }
        }).start();
    }

    private byte[] getBytes(InputStream inputStream) throws Exception {
        ByteArrayOutputStream byteBuffer = new ByteArrayOutputStream();
        int bufferSize = 1024;
        byte[] buffer = new byte[bufferSize];

        int len;
        while ((len = inputStream.read(buffer)) != -1) {
            byteBuffer.write(buffer, 0, len);
        }
        return byteBuffer.toByteArray();
    }
}

File: Permission/app/src/main/java/com/mbsysoft/permissionpractice/MainActivity.kt
package com.mbsysoft.permissionpractice

import android.Manifest
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts

class MainActivity : AppCompatActivity() {

    // 1. activity result launcher 필요
    private val cameraResultLauncher : ActivityResultLauncher<String> =
        registerForActivityResult(
            ActivityResultContracts.RequestPermission()){ //권한 요청
                // 요청이 승인 됐는지 확인
                // 액티비티 결과를 register로 입력하면 activity result launcher가 돌아옴
                isGranted ->
                if(isGranted) {
                    Toast.makeText(this, "퍼미션 완료", Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(this, "퍼미션 거부", Toast.LENGTH_SHORT).show()
                }
            }

    // 2. 퍼미션 여러개 받기
    private val cameraAndLocationResultLauncher : ActivityResultLauncher<Array<String>> =
        registerForActivityResult(
            ActivityResultContracts.RequestMultiplePermissions()){ //권한 요청
            permission ->
            permission.entries.forEach {
                // 리턴 => it:Map.Entry<String, Boolean>
                // 키 : 키워드 벨류 : 승인여부
                val permissionName = it.key
                val isGranted = it.value

                if(isGranted) { //승인 확인
                    if (permissionName == Manifest.permission.ACCESS_FINE_LOCATION) {
                        Toast.makeText(this, "GPS 위치 퍼미션", Toast.LENGTH_SHORT).show()
                    } else if (permissionName == Manifest.permission.ACCESS_COARSE_LOCATION) {
                        Toast.makeText(this, "대략적 위치 퍼미션", Toast.LENGTH_SHORT).show()
                    } else {
                        Toast.makeText(this, "카메라 퍼미션", Toast.LENGTH_SHORT).show()
                    }
                } else {
                    if (permissionName == Manifest.permission.ACCESS_FINE_LOCATION) {
                        Toast.makeText(this, "위치 퍼미션 거부", Toast.LENGTH_SHORT).show()
                    } else if (permissionName == Manifest.permission.ACCESS_COARSE_LOCATION) {
                        Toast.makeText(this, "대략적 위치 퍼미션 거부", Toast.LENGTH_SHORT).show()
                    } else {
                        Toast.makeText(this, "카메라 퍼미션 거부", Toast.LENGTH_SHORT).show()
                    }
                }
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val btnPermission : Button = findViewById(R.id.btnPermission)
//        btnPermission.setOnClickListener {
//            if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
//                && shouldShowRequestPermissionRationale(Manifest.permission.CAMERA)) {
//                Toast.makeText(this, "액세스 거부되어 실행 불가", Toast.LENGTH_SHORT).show()
//            } else {
//                cameraResultLauncher.launch(Manifest.permission.CAMERA)
//            }
//        }

        btnPermission.setOnClickListener {
            if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
                && shouldShowRequestPermissionRationale(Manifest.permission.CAMERA)) {
                Toast.makeText(this, "액세스 거부되어 실행 불가", Toast.LENGTH_SHORT).show()
            } else {
                cameraAndLocationResultLauncher.launch(
                    arrayOf(Manifest.permission.CAMERA,
                        Manifest.permission.ACCESS_FINE_LOCATION,
                        Manifest.permission.ACCESS_COARSE_LOCATION,)
                    )
            }
        }

    }
}